### PR TITLE
Integrate PDAL .las output into VTK applets

### PR DESCRIPTION
**Pull request by willdunklin**

### CODE REVIEW

File: arrows/vtk/CMakeLists.txt [hunk 1]
```diff
@@ -51,6 +51,7 @@ endif()
 
 target_link_libraries( kwiver_algo_vtk
   PUBLIC               vital
+                       kwiver_algo_pdal
```
Comment (mleotta) [inline arrows/vtk/CMakeLists.txt:54]: This should go in the `PRIVATE` section below because the PDAL includes are only used in the `.cxx` file and not exposed in the `.h`.

File: arrows/vtk/applets/fuse_depth.cxx [hunk 2]
```diff
@@ -81,6 +82,10 @@ bool check_config(kv::config_block_sptr config)
     validate_required_input_dir("input_cameras_directory", *config,
                                  main_logger) && config_valid;
 
+  config_valid =
+    validate_required_input_file("input_geo_origin_filename", *config,
```
Comment (mleotta) [inline arrows/vtk/applets/fuse_depth.cxx:86]: This should probably be `validate_optional_input_file` because the geo origin only required if you are also writing a LAS file.  Actually, if you move this check later after checking the file extension below you could have something like:
```c++
if (extension == ".las" )
{
  config_valid =
    validate_required_input_file("input_geo_origin_filename", *config,
                                 main_logger) && config_valid;
}
else
{
  config_valid =
    validate_optional_input_file("input_geo_origin_filename", *config,
                                 main_logger) && config_valid;
}
```

File: arrows/vtk/applets/fuse_depth.cxx [hunk 3]
```diff
@@ -366,10 +382,29 @@ class fuse_depth::priv
         mesh_writer->AddInputDataObject(isosurface_mesh);
         mesh_writer->Write();
       }
+      else if ( extension == ".las" )
+      {
+        auto lgcs = vital::local_geo_cs();
+        read_local_geo_cs_from_file(lgcs, input_geo_origin_file);
```
Comment (mleotta) [inline arrows/vtk/applets/fuse_depth.cxx:388]: We probably want some sort of check here that the `lgcs` is valid.  I don't remember if `read_local_geo_cs_from_file` throws an exception or returns `false` on failure, but it would be good to catch failures somehow so we don't have a repeat of the odd behavior you observed previously with the empty geo origin file.

**Review by mleotta — CHANGES_REQUESTED** [review]
overall looks pretty good, but see inline comments

File: arrows/vtk/applets/fuse_depth.cxx (hunk 3 above)
Comment (willdunklin) [reply arrows/vtk/applets/fuse_depth.cxx:402]: The `read_local_geo_cs_from_file` function doesn't seem to respond at all on failure. The garbage data seemingly comes from [here](https://github.com/Kitware/kwiver/blob/master/vital/types/local_geo_cs.cxx#L46):
```c++
  {
    std::ifstream ifs(file_path);
    double lat, lon, alt;
->  ifs >> lat >> lon >> alt;
    lgcs.set_origin( geo_point( vector_3d(lon, lat, alt), SRID::lat_lon_WGS84) );
  }
```
I could add some semantic in the scope of `fuse_depth.cxx` checks to make sure that the data is within certain bounds 
```c++
  read_local_geo_cs_from_file(lgcs, input_geo_origin_file);
  auto coords = lcgs.origin().location();
  if (coords[0] < ... || coords[1] > ... || ...)
    ...
```
However, shouldn't this be solved in a modification of `read_local_geo_cs_from_file` in tandem with checks in  `write_local_geo_cs_to_file` to only create a geo origin file if the data is valid?

File: arrows/vtk/CMakeLists.txt (hunk 1 above)
Comment (willdunklin) [reply arrows/vtk/CMakeLists.txt:54]: Adding this change this causes the build to fail. Could the build failing have something to do with some detail in `kwiver_algo_pdal`'s [creation](https://github.com/Kitware/kwiver/blob/master/arrows/pdal/CMakeLists.txt#L28)? Or possibly the fact that it is a static library?

Comment (mleotta) [reply arrows/vtk/CMakeLists.txt:54]: Can you post the error?

Comment (willdunklin) [reply arrows/vtk/CMakeLists.txt:54]:
```bash
/usr/bin/ld: CMakeFiles/kwiver_algo_vtk_applets.dir/fuse_depth.cxx.o: in function `kwiver::arrows::vtk::fuse_depth::priv::write_isosurface()':
fuse_depth.cxx:(.text._ZN6kwiver6arrows3vtk10fuse_depth4priv16write_isosurfaceEv[_ZN6kwiver6arrows3vtk10fuse_depth4priv16write_isosurfaceEv]+0x48b): undefined reference to `kwiver::arrows::pdal::pointcloud_io::save_(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, kwiver::vital::local_geo_cs const&, std::vector<Eigen::Matrix<double, 3, 1, 0, 3, 1>, std::allocator<Eigen::Matrix<double, 3, 1, 0, 3, 1> > > const&, std::vector<kwiver::vital::rgb_color, std::allocator<kwiver::vital::rgb_color> > const&)'
collect2: error: ld returned 1 exit status
```

File: arrows/vtk/applets/fuse_depth.cxx (hunk 3 above)
Comment (mleotta) [reply arrows/vtk/applets/fuse_depth.cxx:402]: Yes, it would be better to put these checks in `read_local_geo_cs_from_file`.  As a starter, we should initialize those local `lat, lon, alt` variables to `0.0`.  That is probably the cause of the non-repeatable behavior when an empty file was used.  We can also do something like `bool valid = ifs >> lat >> lon >> alt;` to check if the read was successful and return false or thrown an exception if not.  We should least have a failure/exception if the file is not opened or not read.  We could also consider checking if `lat` and `lon` are in valid ranges, but that's not as critical.  It would be good to add a unit test for this function that tries to load an empty file and test that it fails.

It might be best to make changes to `read_local_geo_cs_from_file` and related tests in a different PR.

File: arrows/vtk/applets/color_mesh.cxx [hunk 4]
```diff
@@ -64,6 +65,31 @@ bool check_config(kv::config_block_sptr config)
...
+        validate_optional_input_file("input_geo_origin_filename", *config,
+                                     main_logger) && config_valid;
+    }
+  }
+  else
+  {
+    config_valid =
+        validate_optional_input_file("input_geo_origin_filename", *config,
+                                     main_logger) && config_valid;
+  }
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:91]: You can simplify the `else` conditions a little like this:

```suggestion
  std::string extension = "";
  if (config->has_value("output_mesh"))
  {
    std::string output_mesh_ =
      config->get_value<std::string>("output_mesh");
    extension = ST::GetFilenameLastExtension( output_mesh_ );
  }
  if (extension == ".las")
  {
    config_valid =
      validate_required_input_file("input_geo_origin_filename", *config,
                                   main_logger) && config_valid;
  }
  else
  {
    config_valid =
      validate_optional_input_file("input_geo_origin_filename", *config,
                                   main_logger) && config_valid;
  }
```

File: arrows/vtk/applets/color_mesh.cxx [hunk 5]
```diff
@@ -64,6 +65,31 @@ bool check_config(kv::config_block_sptr config)
     validate_required_input_file("video_source", *config, main_logger)
     && config_valid;
 
+  if (config->has_value("output_mesh"))
+  {
+    std::string output_mesh_ =
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:70]: style hint:  The `_` suffix is often used to distinguish class member variables from other local variables, so it probably shouldn't be used here.  We are not super consistent about this.  Some code in KWIVER uses a `m_` prefix instead of a `_` suffix for this purpose.

File: arrows/vtk/applets/color_mesh.cxx [hunk 6]
```diff
@@ -398,6 +432,42 @@ class color_mesh::priv : public kwiver::arrows::vtk::mesh_coloration
       writer->Write();
       return true;
     }
+    else if (ext == ".las")
+    {
+      auto lgcs = vital::local_geo_cs();
+      read_local_geo_cs_from_file(lgcs, input_geo_origin_file_);
+
+      vtkSmartPointer<vtkPoints> inPts = mesh->GetPoints();
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:440]: suggestion: we may want to break out most of the rest of this scope into a helper function, maybe `save_mesh_las()` or similar.  That helper can go in this same file, it just helps to make the code more modular and readable.

File: arrows/pdal/pointcloud_io.h [hunk 7]
```diff
@@ -22,12 +22,6 @@ namespace pdal {
...
 public:
-    /// Write landmarks to a file with PDAL provided a geo origin file
-    void
-    save_(vital::path_t const& filename,
-            vital::path_t const& input_geo_origin_file,
-            vital::landmark_map_sptr const& landmarks);
-
     /// Write landmarks to a file with PDAL
     void
     save_(vital::path_t const& filename,
```
Comment (mleotta) [inline arrows/pdal/pointcloud_io.h:27]: I didn't catch this before, but is there probably isn't a good reason for these functions to belong to a class.  They could just be stand alone functions.  There is no state that the class needs to maintain. 

Also we could just call these functions `save_landmarks_las` or `save_point_cloud_las` since they are specifically writing these data types to a LAS file.  The underscore suffix was used in image_io to distinuish between the pubic `save` function that is exposed and the private `save_` function that developer had to actually implement.  We don't need the underscore suffix here.

File: arrows/pdal/tests/test_pointcloud_io.cxx [hunk 8]
```diff
@@ -50,6 +50,9 @@ TEST_F(pointcloud_io, save_geo_origin) {
     kv::landmark_map_sptr landmark_map =
         kv::read_ply_file(landmarks_path);
 
+    auto lgcs = kwiver::vital::local_geo_cs();
+    read_local_geo_cs_from_file(lgcs, geo_origin_path);
+
     auto io = new kwiver::arrows::pdal::pointcloud_io();
```
Comment (mleotta) [inline arrows/pdal/tests/test_pointcloud_io.cxx:56]: This is a memory leak if there is no corresponding `delete` call.  However, you don't need this line at all if you follow the suggestion above to just use stand alone functions.

File: arrows/vtk/applets/CMakeLists.txt [hunk 9]
```diff
@@ -43,3 +43,7 @@ kwiver_add_plugin( kwiver_algo_vtk_applets
                kwiversys
                ${VTK_targets}
   )
+
+target_link_libraries( kwiver_algo_vtk_applets
+  PRIVATE              kwiver_algo_pdal
+  )
```
Comment (mleotta) [inline arrows/vtk/applets/CMakeLists.txt:49]: I think you can put this library right in the list above in `kwiver_add_plugin`

File: arrows/vtk/applets/color_mesh.cxx [hunk 10]
```diff
@@ -398,13 +427,56 @@ class color_mesh::priv : public kwiver::arrows::vtk::mesh_coloration
...
+      if (rgbArray)
+      {
+        vtkIdType idx = 3 * i;
+        colors[i] = vital::rgb_color(rgbArray->GetValue(idx),
+                                     rgbArray->GetValue(idx + 1),
+                                     rgbArray->GetValue(idx + 2));
+      }
+    }
+
+    auto pointcloud_writer = new kwiver::arrows::pdal::pointcloud_io();
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:476]: again, memory leak, but allocation not needed if you refactor as described above.

File: arrows/vtk/applets/fuse_depth.cxx [hunk 11]
```diff
@@ -366,10 +396,29 @@ class fuse_depth::priv
...
+
+        std::vector<vital::vector_3d> points(numPts);
+        std::vector<vital::rgb_color> colors;
+
+        for (vtkIdType i = 0; i < numPts; ++i)
+        {
+          inPts->GetPoint(i, points[i].data());
+        }
+
+        auto pointcloud_writer = new kwiver::arrows::pdal::pointcloud_io();
```
Comment (mleotta) [inline arrows/vtk/applets/fuse_depth.cxx:415]: again, memory leak, but allocation not needed if you refactor as described above.

File: arrows/vtk/applets/CMakeLists.txt (hunk 9 above)
Comment (mleotta) [reply arrows/vtk/applets/CMakeLists.txt:49]: Actually, we may want to do it this way but wrap this in a conditional `if( KWIVER_ENABLE_PDAL )`.  We should move all that conditional logic with cmakedefines and ifdefs to here from the PDAL arrow.  In the PDAL itself we don't need that logic because either we build the whole arrow or we don't depending on whether PDAL is enabled.  However, we want to build the VTK arrow whether or not PDAL is enabled and only add these PDAL dependencies if PDAL is also enabled.  Does that make sense?

File: arrows/vtk/applets/color_mesh.cxx [hunk 12]
```diff
@@ -5,6 +5,7 @@
 #include "color_mesh.h"
 
 #include <arrows/core/colorize.h>
+#include <arrows/pdal/pointcloud_io.h>
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:8]: This should be in an ifdef checking if PDAL is enabled.

File: arrows/vtk/applets/color_mesh.cxx [hunk 13]
```diff
@@ -398,13 +427,56 @@ class color_mesh::priv : public kwiver::arrows::vtk::mesh_coloration
...
+      return true;
+    }
     else
     {
       LOG_ERROR(main_logger, "Invalid file format: " << ext);
       return false;
     }
   }
 
+  void save_mesh_las(vtkSmartPointer<vtkPolyData> mesh,
```
Comment (mleotta) [inline arrows/vtk/applets/color_mesh.cxx:442]: This function body should be in an ifdef checking if PDAL is enabled.  Thrown an exception or return false if it's not enabled.

File: arrows/vtk/applets/fuse_depth.cxx [hunk 14]
```diff
@@ -6,6 +6,7 @@
 
 #include <arrows/core/depth_utils.h>
 #include <arrows/mvg/sfm_utils.h>
+#include <arrows/pdal/pointcloud_io.h>
```
Comment (mleotta) [inline arrows/vtk/applets/fuse_depth.cxx:9]: This should be in an ifdef checking if PDAL is enabled.

File: arrows/vtk/applets/fuse_depth.cxx (hunk 11 above)
Comment (mleotta) [inline arrows/vtk/applets/fuse_depth.cxx:415]: This code should be in an ifdef checking if PDAL is enabled. Log and error and return false if it's not enabled.

**Review by mleotta — CHANGES_REQUESTED** [review]
See inline comments.

File: arrows/vtk/applets/CMakeLists.txt (hunk 9 above)
Comment (willdunklin) [reply arrows/vtk/applets/CMakeLists.txt]: Makes sense to me. In my implementation I left the `pdal_config.h.in` file in the pdal arrow. Should I move that into the vtk applets as well?

File: CMake/kwiver_config.h.in [hunk 15]
```diff
@@ -0,0 +1,17 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/**
```
Comment (mleotta) [inline CMake/kwiver_config.h.in:5]: I think according to our latest style guide, we not want to avoid `/**` style doxygen documentation and use only `///` style instead.

File: arrows/pdal/pointcloud_io.cxx [hunk 16]
```diff
@@ -8,39 +8,25 @@
...
-  auto lgcs = vital::local_geo_cs();
-  read_local_geo_cs_from_file(lgcs, input_geo_origin_file);
-  pointcloud_io::save_(filename, lgcs, landmarks);
-}
-
 /// Write landmarks to a file with PDAL
 void
-pointcloud_io::save_(vital::path_t const& filename,
+save_point_cloud_las(vital::path_t const& filename,
              vital::local_geo_cs const& lgcs,
```
Comment (mleotta) [inline arrows/pdal/pointcloud_io.cxx:30]: Minor formatting nit pick:  If if doesn't make the line lengths too long, I would indent the second and following lines so that the variable types line up with the first line.   The current indention looks arbitrary.  So do this

```c++
save_point_cloud_las(vital::path_t const& filename,
                     vital::local_geo_cs const& lgcs,
                     vital::landmark_map_sptr const& landmarks)
```
or this
```c++
save_point_cloud_las(
  vital::path_t const& filename,
  vital::local_geo_cs const& lgcs,
  vital::landmark_map_sptr const& landmarks)
```

They same formatting rule applies everywhere else you are declaring or defining functions.

**Review by mleotta — CHANGES_REQUESTED** [review]
This looks good now other than some very minor formatting issue.

**Review by mleotta — APPROVED** [review]

**Comment by mleotta** [comment]
jenkins test this please
